common/hashing_benchmark.cpp

```diff
@@ -0,0 +1,251 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <benchmark/benchmark.h>
+
+#include <algorithm>
+#include <cstddef>
+
+#include "absl/hash/hash.h"
+#include "absl/random/random.h"
+#include "common/hashing.h"
+#include "llvm/ADT/Hashing.h"
+
+namespace Carbon {
+namespace {
+
+// We want the benchmark working set to fit in the L1 cache where possible so
+// that the benchmark focuses on the CPU-execution costs and not memory latency.
+// For most CPUs we're going to care about, 16k will fit easily, and 32k will
+// probably fit. But we also need to include sizes for string benchmarks. This
+// targets 8k of entropy with each object up to 8k of size for a total of 16k.
+constexpr int EntropySize = 8 << 10;
+constexpr int EntropyObjSize = 8 << 10;
+
+// An array of random entropy with `EntropySize` bytes plus 8k. The goal is that
+// clients can read `EntropySize` objects of up to 8k size out of this pool by
+// starting at different byte offsets.
+static const llvm::ArrayRef<std::byte> entropy_bytes =
+    []() -> llvm::ArrayRef<std::byte> {
+  static llvm::SmallVector<std::byte> bytes;
+  // Pad out the entropy for up to 1kb objects.
+  bytes.resize(EntropySize + EntropyObjSize);
+  absl::BitGen gen;
+  for (std::byte& b : bytes) {
+    b = static_cast<std::byte>(absl::Uniform<uint8_t>(gen));
+  }
+  return bytes;
+}();
```

Comment:
Driveby. Would it help with Josh's concern about noticing the invocation to use something like `std::invoke` so the fact that it's being invoked is upfront rather than relegated to the easily overlooked `()`?

Reply:
I'd be open to that option. In general I find that the toolchain code is employing a lot of idioms that C++ experts are comfortable with but I suspect are going to be a barrier to most other contributors, and I'd like to make an effort to reign it in where it is delivering only marginal benefit. `std::invoke` at least has the benefit of being something you could do a web search for.

Reply:
I've added a section to [the idioms in the toolchain doc](https://docs.google.com/document/d/1RRYMm42osyqhI2LyjrjockYCutQ5dOf8Abu50kTrkX0/edit?resourcekey=0-kHyqOESbOHmzZphUbtLrTw#heading=h.cttvi8w0v7e). Please update this as well once the code is updated.